feat(workflow): Support `base` collapse and new endpoint to fetch stats

iProgramStuff opened the pull request:
This PR supports our efforts towards splitting up the issue stats into multiple requests. It accomplishes this in two ways:

1. It adds collapse=base support to the group serializer, so that we can use all of its existing functionality (Was going to have a separate serializer at first but this seemed like the better path and gives us more options) of that serializer and the current endpoint - if, for some reason, we ever wanted to just get stats but with the search functionality of the current endpoint.

2. It adds a new endpoint, much like the existing one except that it takes a list of groups and defaults to collapsing the base stats if collapse is not passed. The new endpoint is at `/issues-stats/` (the original is `/issues/`).

So we'll now be able to call the original endpoint with `collapse=stats` and their search query to get the initial results to display, and then request the stats for individual groups (all at once or in batches as we experiment).

iProgramStuff committed:
Adding json schema for GroupInbox reason_details

iProgramStuff committed:
Support collapse

iProgramStuff committed:
Oops.

iProgramStuff committed:
First draft of new stats endpoint + base collapse support

iProgramStuff committed:
Merging with master

iProgramStuff committed:
Letting detail props be null

iProgramStuff committed:
Merge remote-tracking branch 'origin' into feat/jsonschemagroupinbox

iProgramStuff committed:
Fixng JSON schema bug and slipping in SENTRY-JG9 fix

iProgramStuff committed:
Fixng JSON schema bug and slipping in SENTRY-JG9 fix

iProgramStuff committed:
Merging with master

iProgramStuff committed:
Fixing tests

iProgramStuff committed:
Cleaning up some todos

iProgramStuff committed:
Updating comment

iProgramStuff committed:
Adding inbox feature to test

markstory commented on src/sentry/api/endpoints/organization_group_index_stats.py:
```diff
@@ -0,0 +1,127 @@
+            return Response({"detail": "Group ids must be integers"}, status=400)
+
+        if not group_ids:
+            return Response(
```
You could use `rest_framework.exceptions.ParseError` to make these error responses a bit more terse.

markstory commented on src/sentry/api/endpoints/organization_group_index_stats.py:
```diff
@@ -0,0 +1,127 @@
+            stats_period_end = end
+        else:
+            stats_period_start = None
+            stats_period_end = None
```
Could this logic be shared with any of the other group endpoints? Seems like there'd be more copies of it about.

iProgramStuff replied:
Yeah, will definitely try to reuse this logic a bit more. Straight copy pasta'd this from the other endpoint.

iProgramStuff committed:
Centralizing stats calculation

iProgramStuff committed:
Merge remote-tracking branch 'origin' into feat/groupstatsserializer

iProgramStuff committed:
ParseError & Merge with master

evanpurkhiser committed:
ref(stats): Link to project issues, not settings (#21703)

jan-auer committed:
fix(issue-details): Switch labels for absolute/relative addrs toggle (#22098)

The labels for toggling between absolute and relative addresses are
swapped. This fixes the order, showing "Switch to absolute" when
relative addresses are shown, and vice versa.

iProgramStuff committed:
Merging with branch for bugfix

iProgramStuff committed:
Merge remote-tracking branch 'origin' into feat/jsonschemagroupinbox

iProgramStuff committed:
Merge branch 'feat/jsonschemagroupinbox' into feat/groupstatsserializer

MeredithAnya commented on src/sentry/api/endpoints/organization_group_index.py:
```diff
@@ -101,7 +101,8 @@ def get(self, request, organization):
                                           issues belong to.
         :auth: required
-        TODO(Chris F.): Add details on expand/collapse.
+        :qparam list expand an optional list of strings to opt in to additional data. Supports `inbox`
+        :qparam list collapse an optional list of strings to opt out of certain pieces of data. Supports `stats`, `lifetime`, `base`
```
yeah @iProgramStuff you'll have to update the schema for this endpoint https://github.com/getsentry/sentry/tree/master/api-docs/paths

mikellykels committed:
style(ui): Rename Merged to Merged Issues (#22069)

Rename Merged to Merged Issues on Issue page.

evanpurkhiser committed:
ref(py3): Fix keys error in legacy jira plugin (#22109)

Fixes: SENTRY-JJA

scttcper committed:
fix(workflow): Revert inbox backlog query, revert default query (#22106)

* fix(workflow): Revert inbox backlog query, revert default query

* rename backlog tab to unresolved

evanpurkhiser committed:
fix(p3): Fix byte joining in old jira client (#22110)

scttcper committed:
fix(ui): Fix organization context remount binding (#22112)

ckj committed:
ref(ui): use prism-sentry theme for code snippets (#22095)

evanpurkhiser committed:
ref(ts): Convert platformPicker (#21708)

wedamija commented on src/sentry/api/endpoints/organization_group_index.py:
```diff
@@ -235,7 +226,7 @@ def get(self, request, organization):
         ]
         if status and status[0].value.raw_value == GroupStatus.UNRESOLVED:
-            context = [r for r in context if r["status"] == "unresolved"]
+            context = [r for r in context if "status" not in r or r["status"] == "unresolved"]
```
Why do we need this?

iProgramStuff replied:
If we call this endpoint with `collapse=base`, status won't be included and this would cause errors

wedamija commented on src/sentry/api/endpoints/organization_group_index_stats.py:
```diff
@@ -0,0 +1,118 @@
+        context = serialize(
+            groups,
+            request.user,
+            StreamGroupSerializerSnuba(
```
Why do we serialize the whole group here? Won't we have fetched them from the old endpoint?

iProgramStuff replied:
We're skipping the all of the work done in the core serializer when we call this with `collapse=base`.

wedamija commented on src/sentry/api/endpoints/organization_group_index.py:
```diff
@@ -102,7 +103,8 @@ def get(self, request, organization):
                                           issues belong to.
         :auth: required
-        TODO(Chris F.): Add details on expand/collapse.
+        :qparam list expand an optional list of strings to opt in to additional data. Supports `inbox`
+        :qparam list collapse an optional list of strings to opt out of certain pieces of data. Supports `stats`, `lifetime`, `base`
```
Should we be passing these to the serialize function?

priscilawebdev committed:
ref(ts): Convert accountSecurityDetails to typescript (#22026)

matejminar committed:
feat(ui): Init project detail page (#22054)

priscilawebdev committed:
ref(group-list): Improvements + add useFilteredStats prop in the StreamGroup Component (#22084)

priscilawebdev committed:
fix(ui-group-list): Fix 't' is undefined - JAVASCRIPT-238M

priscilawebdev committed:
fix(ui): Fix groupList issue (#22127)

armenzg committed:
fix(issues-ui) - MultiEnvironmentSelector to show environments to superuser (#22090)

Superusers accessing a customer's organization do not belong to any specific team.
Nevertheless, the superuser should not only see all projects but all environments
associated to those projects.

This change fixes the above.

Fixes ISSUE-860

Co-authored-by: Armen Zambrano Gasparnian <armenzg@Armens-MacBook-Pro.local>

markstory committed:
ref(ts) Convert AllTeamsRow to typescript (#22093)

Zylphrex committed:
fix(releases): Issues dropdown should render above transactions dropdown button (#22105)

Currently the issues dropdown options render below the transactions dropdown
button because of z-index. This change ensures that it renders above the button.

markstory committed:
ref(ts) Convert team details wrapper to typescript (#22102)

* ref(ts) Convert team details wrapper to typescript

Replace `createReactClass()` and mixins with the `withTeams` HOC.

* Remove else

priscilawebdev committed:
fix(getThreadException): Fix edge case issue (#22052)

priscilawebdev committed:
ref(ts): Convert PasswordForm to Typescript (#22121)

priscilawebdev committed:
fix(rawStacktraceContent): Fix cannot read frames of null (#22126)

markstory committed:
feat(dashboards) Adapt API to new behavior round 2 (#22092)

This reverts commit 4fdeba2522f422e04a88f1ba66ea051e36cbd37c.

This change restores the changes to dashboards and shouldn't break
getsentry tests as the old models are retained in django but have their
foreign keys removed which will unblock truncate operations.

We're reviving custom dashboards, and need to adapt them to work with
discover2 style queries.

* Rename Widget to DashboardWidget as Widget is a common term and using
  it here unprefixed will make future us sad.
* Rename WidgetDataSource to DashboardWidgetQuery. There is no
  flexibility in widget datasources now. They will all use discover.
* Hide order attributes for widgets and queries. The order in the API
  request/response is the implict order. We don't need additional API
  parameters.
* Add `interval` to queries so that users can pick bin sizes for their
  charts.
* Use the order of the elements in the list instead of an additional order
   parameter.

There is still some work to be done around validating widgets and their
queries but I wanted to do those separately.

These endpoints are unused in production and should be safe to change
without consequence.

Remove foreign key from widget. This constraint previously broke builds 
in getsentry as truncate
couldn't go.

BYK committed:
meta: Bump new development version

taylangocmen committed:
ref(ts): Convert apiChart to Typescript (#22103)

Commit:
fix(integrations): only throw error if identity is required (#22108)

billyvg committed:
build(ci): Fix merge base for visual snapshots (#22111)

When we moved from the workflow trigger event `pull_request` to `pull_request_target` (which we changed to allow forks to run Visual Snapshots), it resulted in a change of the pull request base SHA. Instead of relying on the github context for the merge base, we should use `git merge-base` to find the base sha.

The merge base SHA is then used to use a three-way comparison for PRs (comparing the PR head SHA against latest master and the merge base).

Also of note, `pull_request_target` is annoying to test - you need to create a PR with your feature branch as the base (because it will only use the workflow from the base branch/master). Here is my testing branch https://github.com/getsentry/sentry/pull/22101 which uses a fork that is quite outdated (around a week or more). The number of snapshot changes is misleading as it doesn't quite detect the change of width as not important. But the tests prefix had 471 changes (vs the 150 or so here).


Fixes https://app.asana.com/0/1197393184799586/1199078519540749/f

evanpurkhiser committed:
ref(py3): Patch compat for pickle.{Unp,P}ickler (#22114)

billyvg committed:
feat(ui): Enable dark mode auto-switching for staff (#22135)

This enables dark mode auto-switching for staff so that they do not need to manually toggle it on.

Commit:
fix(py3): Jira plugin string handling (#22115)

wmak committed:
feat(release-trends): Adding trends to the transaction list (#22012)

- This adds two trends options to the release's transaction list, Trending Regressions & Trending Improvements
- Clicking the transaction name will bring you to the transaction's summary view with the trend display mode already selected
- This also adds that when clicking on releases in trends, the transactions list will already be on the related list

davidenwang committed:
style(metric_alerts): Query field fixed width (#22133)

Added an ability to specify fixed width on the <MetricField> in the metric alert builder. This eliminates unnecessary whitespace in the metric selector and makes the sentence {function} over {timewindow} easier to read.

k-fish committed:
feat(flag): Add feature flags for performance landing changes (#22134)

This adds two feature flags, one for the upcoming performance landing v2, and one for the interim changes to add vitals overview.

scttcper committed:
feat(workflow): Return inbox details on group unresolve (#22074)

markstory committed:
fix(ownership) Fix rules being dropped (#22139)

When I removed the duplicate state I neglected to coalesce the
initialText prop with state. This caused rules to be dropped if the add
button was clicked before any changes were made in the text box.

scttcper committed:
perf: Add collapse latestDeploys flag to projects endpoint (#22116)

k-fish committed:
fix(perf): Fix transaction trends display legend (#22000)

A recent refactor caused the legend to no longer be selectable as the query key is separate for trends to not conflict with other displays

taylangocmen committed:
fix(workflow): Inbox html event annotation tag to set innerhtml (#22136)

billyvg committed:
build(ci): Rename acceptance py 3.6 workflow (#22142)

Rename this workflow to differentiate from the 2.7 acceptance workflow. The job name is going to remain the same to be consistent w/ 2.7 jobs.

billyvg committed:
build(ci): Add/fix webpack instrumentation for GHA (#22034)

This updates our webpack instrumentation so that it works on GHA

davidenwang committed:
feat(metric_alerts): Use in-field labels for query field selector (#22130)

For consistency on the metric alert builder, changed the type of labels from an aggregate header on the metric selector, to an in-field label.

wedamija reviewed (approved):
I'm ok with this, at least for the moment. I think the way I envisioned this would be that once we collapse parameters, we'd fetch them from an entirely separate endpoint. So there might be a stats endpoint that just returns a dict of `{<group_id>: <stats>}`, or something along those lines. 

We can keep iterating on this though, I see the benefit of doing it this way as well.

matejminar committed:
feat: Add ScoreCard to project detail (#22117)

matejminar committed:
ref(ui): Consolidate Create Alert buttons (#22122)

matejminar committed:
feat(ui): Add team access to project detail (#22118)

markstory commented on src/sentry/api/endpoints/organization_group_index_stats.py:
```diff
@@ -0,0 +1,118 @@
+            if not groups:
+                return Response({"detail": "No matching groups found"}, status=400)
+            elif len(groups) > 25:
+                return Response({"detail": "Too many groups requested."}, status=400)
```
```suggestion
                return ParseError(detail="Too many groups requested.")
```

markstory commented on src/sentry/api/endpoints/organization_group_index_stats.py:
```diff
@@ -0,0 +1,118 @@
+        else:
+            groups = list(Group.objects.filter(id__in=group_ids, project_id__in=project_ids))
+            if not groups:
+                return Response({"detail": "No matching groups found"}, status=400)
```
```suggestion
                return ParseError(detail="No matching groups found")
```

wmak committed:
fix(discover): Don't show more than 9 on toggling (#22137)

- This fixes a bug where toggling show prebuilt would keep the pervious
  state, so would either show less than 9 or more than 9 depending on
  which state you were moving to

matejminar committed:
test(acceptance): Add project detail visual snapshots (#22151)

Zylphrex committed:
feat(releases): Add failure count and transaction count graphs (#22091)

This adds a failure count and transaction count graphs to the release details page.

Zylphrex committed:
feat(perf-view): Lock down transaction summary and vitals to a single project (#22138)

When on the transaction summary and transaction vitals page, the global
selection header currently allows the user to change projects. This always
almost leads to a broken page as the same transaction name almost never
translate across projects, leading to broken states. This change locks down
the global selection header so users cannot change the project.

iProgramStuff committed:
Update src/sentry/api/endpoints/organization_group_index_stats.py

Co-authored-by: Mark Story <mark@sentry.io>

iProgramStuff committed:
Update src/sentry/api/endpoints/organization_group_index_stats.py

Co-authored-by: Mark Story <mark@sentry.io>

k-fish committed:
ref(ts): Convert linechart (again) (#22144)

Fixing revert #21823 which started failing in master despite PR passing checks (likely merge issues)

iProgramStuff committed:
feat(workflow): Adding json schema for GroupInbox reason_details (#21882)

* Adding json schema for GroupInbox reason_details

* Slipping in SENTRY-JG9 fix

iProgramStuff committed:
Merge remote-tracking branch 'origin' into feat/groupstatsserializer

iProgramStuff committed:
Adding colon

iProgramStuff committed:
...

iProgramStuff committed:
Fixing raise/return bug
